### PR TITLE
Add a cgroup repo/project proposal

### PR DESCRIPTION
**Pull request by kolyshkin**
This adds a proposal to split github.com/opencontainers/runc/libcontainer/cgroups packages to a new home at github.com/opencontainers/cgroups, which appeared as a result of discussion at https://github.com/kubernetes/kubernetes/issues/128157 (and https://github.com/kubernetes/kubernetes/issues/128157#issuecomment-2427394891 in particular).

Below is a copy-paste of the current draft.

----
# OCI cgroups project proposal

## Abstract
Having a common cgroups management implementation in OCI for use in and outside
of runc will ensure long lasting security and interoperability throughout the
container ecosystem.

## Proposal

Refactor and move cgroups packages out of runc into a separate project:

https://github.com/opencontainers/runc/tree/master/libcontainer/cgroups

The new project would live in the opencontainers github organization:

https://github.com/opencontainers/cgroups

A sample of how the project would look like is already here:

https://github.com/kolyshkin/opencontainers-cgroups

### Initial Maintainers
Initial maintainers of the cgroups project would be:

* Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp> (@AkihiroSuda)
* Aleksa Sarai <cyphar@cyphar.com> (@cyphar)
* Kir Kolyshkin <kolyshkin@gmail.com> (@kolyshkin)
* Mrunal Patel <mpatel@redhat.com> (@mrunalp)
* Sebastiaan van Stijn <github@gone.nl> (@thaJeztah)
* Odin Ugedal <odin@uged.al> (@odinuge)
* Peter Hunt <pehunt@redhat.com> (@haircommander)
* Davanum Srinivas <davanum@gmail.com> (@dims)

### Code of Conduct
This project would incorporate (by reference) the OCI [Code of Conduct][code-of-conduct].

### Governance and Releases
This project would incorporate the Governance and Releases processes from the
OCI project template: https://github.com/opencontainers/project-template.

### Project Communications
The proposed project would continue to use existing channels in use by the OCI
developer community for communication including:
* GitHub for issues and pull requests
* The dev@opencontainers.org email list
* The weekly OCI developer community conference call
* The #OpenContainers IRC channel

### Versioning / Roadmap

We'll start with v0.x until everything is stabilized after the split.

The plan is to have frequent releases to accommodate various users.
## Frequently Asked Questions (FAQ)
*Q: Does this change the OCI Charter or Scope Table?*
A: No. Nothing in this proposal is intended to amend the
[OCI Charter](https://www.opencontainers.org/about/governance) or
[OCI Scope Table](https://www.opencontainers.org/about/oci-scope-table).

*Q: Why move this out of the runc project?*
A: To be able to reuse this in different container projects, while avoiding
code duplication and fragmentation. This involves adding functionality and
cutting releases to accomodate existing users other than runc.

Here's an example: kubernetes needs PSI stats from cgroups, and while the
low-level functionality needed ([runc PR #3900](https://github.com/opencontainers/runc/pull/3900))
was merged in July 2023, it was only made available to kubernetes when runc
v1.2.0 was tagged in October 2024. Such delays might result in using
unversioned dependencies, or copying the code around and thus fragmentation.

A longer term goal is make this set of packages even more modular, avoiding
excessive third-party dependencies.

*Q: Who are the other target users of these packages?*

A: CRI-O, kubernetes, cadvisor.

[code-of-conduct]: https://github.com/opencontainers/org/blob/master/CODE_OF_CONDUCT.md

<hr>

In order for this vote to pass, at least two thirds of the TOB (6 members) must vote in favour of this proposal:

 - [x] Brandon Mitchell @sudo-bmitch 
 - [x] Vincent Batts @vbatts 
 - [x] Giuseppe Scrivano @giuseppe 
 - [x] Phil Estes @estesp 
 - [ ] Ramkumar Chinchani @rchincha 
 - [ ] Sajay Antony @sajayantony 
 - [x] Samuel Karp @samuelkarp 
 - [x] Derek McGowan @dmcgowan 
 - [x] Aleksa Sarai @cyphar

### CODE REVIEW

**Comment by samuelkarp** [comment]
New projects require a 2/3 vote of the TOB.  The TOB has 9 seats, so this means 6/9 must approve.

@sudo-bmitch @vbatts @giuseppe @estesp @rchincha @sajayantony @samuelkarp @dmcgowan @cyphar

**Comment by samuelkarp** [comment]
> TODO add someone from k8s org

Are there suggestions for this?  Or should we remove for now and add later through the project's own governance?

**Comment by AkihiroSuda** [comment]
- What was the problem with http://github.com/containerd/cgroups ?
- Can we just split the Go module without splitting the repo, for ease of contribution and running CI?

**Comment by kolyshkin** [comment]
> > TODO add someone from k8s org
> 
> Are there suggestions for this? Or should we remove for now and add later through the project's own governance?

I was hoping we can discuss it right here in this PR and add someone before we merge it.
Alas I'm not very familiar with k8s people; maybe @dims, @liggitt or @odinuge will volunteer.

I guess we also need LGTMs from all the to-be-maintainers (whom I've listed boldly without even talking to). Those are:
 * [ ] @AkihiroSuda
 * [ ] @cyphar 
 * [x] @kolyshkin
 * [ ] @mrunalp
 * [ ] @thaJeztah
Please let us know here if you can keep maintaining what used to be opencontainers/runc/libcontainer/cgroups.

**Comment by kolyshkin** [comment]
> * What was the problem with [http://github.com/containerd/cgroups](https://github.com/containerd/cgroups) ?

* For kubernetes, see https://github.com/kubernetes/kubernetes/issues/128157.
* For runc, I imagine it lacks a lot of stuff needed (device management is one big thing) and in general is less stable/tested/mature.

> * Can we just split the Go module without splitting the repo, for ease of contribution and running CI?

Possibly, but I don't really see the point. To me it looks like
 - contribution is going to be mostly the same, except maybe some new cgroup settings will require two PRs into two repos -- but it's good because we'll have some k8s people look at the changes.
 - CI is going to be more complicated with a monorepo (need to add `cd libcontainers/cgroups && go test|mod|etc` everywhere).

**Comment by AkihiroSuda** [comment]
How will we test PRs for `opencontainers/cgroups`?
Many regressions cannot be exposed without running the whole runc CI for several distributions. (And sometimes it is not even enough)

**Comment by dims** [comment]
I am happy to!

**Comment by mrunalp** [comment]
LGTM

**Comment by kolyshkin** [comment]
> How will we test PRs for `opencontainers/cgroups`?
> Many regressions cannot be exposed without running the whole runc CI for several distributions. (And sometimes it is not even enough)

We can obviously test opencontainers/cgroups HEAD in runc repo, but that will only cover code which is already merged.

Good question. The best thing I can come up with is opening a PR in cgroups will result in opening a (linked) PR in a (hypothetical) runc-test repo (which is synced to main runc repo) with the libct/cg code vendored from the original PR. That is too complicated, I agree.

So, how do we currently deal with a case when someone makes a change in one of the packages that runc uses? Let's say it's a  change in `github.com/coreos/go-systemd` or `github.com/opencontainers/selinux` or `github.com/urfave/cli`. Let's say the author tests their changes in runc repo. If they do, they will catch a bug early. If they don't, and there is a bug, we'll catch it when a new release is tagged, and we roll back to using a previous release and fix the bug in the next one. I suggest we deal with opencontainers/cgroups in the same way.

The alternative, I guess, is to keep cgroups in the same repo, add `go.mod`/`go.sum` to libcontainer/cgroups directory in runc, and make releases from there (marking them with `libcontainer/cgroups/vX.Y.Z` tag). This way we also have a problem (besides some extra releases and tags in runc repo), or should I say a dilemma -- whether to use a tagged version of cgroups for runc, or use a latest version (via `replace github.com/opencontainers/runc/libcontainer/cgroups => ./libcontainer/cgroups` directive in runc's `go.mod`). Problem is, we can not use both (or we need to double our CI jobs).

So, to my mind, it's better to treat cgroups as yet another package (set of packages) which is used by runc, and deal with it accordingly (fix bugs if/when found, test some major changes in both repos etc.)

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by sudo-bmitch** [comment]
In addition to verifying the new maintainers are on-board, I'd like to hear from the existing @opencontainers/runc-maintainers.

- [X] @mrunalp
- [ ] @dqminh
- [ ] @hqhq
- [X] @cyphar
- [ ] @AkihiroSuda
- [X] @kolyshkin
- [ ] @thaJeztah
- [ ] @lifubang
- [ ] @rata

**Review by rata — APPROVED** [review]
LGTM.

Regarding the concerns that @AkihiroSuda raised, quite good questions, I think what @kolyshkin suggested seems good to start. Having a quick look, is not the most active place in runc the libcontainers/cgroups:

```
runc/libcontainer/cgroups :main$ git log . | grep ^Date| grep 2024 | wc -l
17
runc/libcontainer/cgroups :main$ git log . | grep ^Date| grep 2023 | wc -l
44
runc/libcontainer/cgroups :main$ git log . | grep ^Date| grep 2022 | wc -l
27
runc/libcontainer/cgroups :main$ cd ../../
runc :main$ git log . | grep ^Date| grep 2022 | wc -l
310
runc :main$ git log . | grep ^Date| grep 2023 | wc -l
426
runc :main$ git log . | grep ^Date| grep 2024 | wc -l
286
```

Doing this will clearly improve the current situation, if it ends up causing some other issues, when we know the actual issues we can think how to solve them.

The bringing it back to the runc repo and use a different go.mod, etc. might be an option. But it has many downsides too, I wouldn't want to start with that. I think the new repo is the best option to try now.

**Comment by odinuge** [comment]
Big +1 on this @kolyshkin, thanks for pushing. I'm also a big fan of this proposal instead of a fork.

I've also been working here a bit from the k8s side, and I'm happy to volunteer. So feel free to tag me in for reviews. I've poked at a lot of this code, and things like the systemd integration is very brittle, so having both people from k8s and the runc/opencontainer communities work together seems like a good idea here.

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
Updated the proposal; here's the diff:
```diff
-* TODO add someone from k8s org
+* Odin Ugedal <odin@uged.al> (@odinuge)
+* Peter Hunt <pehunt@redhat.com> (@haircommander)
+* Davanum Srinivas <davanum@gmail.com> (@dims)
 ```

**Review by lifubang — APPROVED** [review]
LGTM
Cgroup is one of the most important core dependencies of runc, so I have a strong suggestion to add a runc(or containerd) integration test in CI to test PRs, but this may cause CI fail if the PR breaks some APIs or aims to fix a bug. Maybe we should have our own integration tests in the new repo. I think the time spent on CI is worth it.

**Review by hqhq — APPROVED** [review]

**Comment by sudo-bmitch** [comment]
@thaJeztah are you okay with being included as a moderator on this?

@kolyshkin thank you for adding the k8s maintainers. Are there any other pending changes to this before the TOB should vote on it? (We vote on the commit digest.)

**Comment by kolyshkin** [comment]
> @kolyshkin thank you for adding the k8s maintainers. Are there any other pending changes to this before the TOB should vote on it? (We vote on the commit digest.)

I do not have anything to add, thanks!

**Comment by thaJeztah** [comment]
Sorry, I'm a bit slow on my notifications.

Generally, LGTM; for the testing part; yes, there may be some situations where it's slightly more complicated to verify changes; that said, indeed it's possible to add a `replace` rule in runc if someone is working on changes in cgroups that have corresponding changes in runc. To some extent, I consider the extra hurdle to take a _benefit_; it makes anyone making changes more aware if they're making changes to a public API (which _should_ always have some friction :smile:). 

As to merging with containerd/cgroups; I don't think that needs to be a blocker for this effort; starting with a v0.x.x tag still allows for some flexibility on defining the public / stable API (we should definitely have a good look at what parts we're comfortable with being "public"), and we can maintain aliases in the old location for some time to help the transition. Definitely would be great if we could at some point consolidate the implementations (I don't think anyone _really_ benefits from having 2 separate projects for this).

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by sudo-bmitch — APPROVED** [review]
LGTM

**Review by dmcgowan — APPROVED** [review]

**Review by samuelkarp — APPROVED** [review]
LGTM

**Review by estesp — APPROVED** [review]

**Review by vbatts — APPROVED** [review]
LGTM

**Review by cyphar — APPROVED** [review]

**Comment by utam0k** [comment]
Is the scope of this proposal only in the Go language? I'm looking for the possibility of supporting Rust as well.

**Comment by sudo-bmitch** [comment]
> Is the scope of this proposal only in the Go language? I'm looking for the possibility of supporting Rust as well.

The runc folder linked above is only Go code. I think rust would be a separate request, and would require enough maintainers that want to manage it to step up.

**Comment by cyphar** [comment]
@utam0k While it would be nice for us to have less of a Go-only focus in OCI, I do wonder what a user would expect from a Rust and Go implementation of some library functionality. Would they want feature parity? If so, there's a lot of ugly stuff you would need to reimplement in Rust that is mostly legacy stuff borne out of runc's history -- though hopefully we can remove some of that when we make cgroups a separate module.

How usable (as a library) do you feel is the cgroup code in youki?

**Comment by sudo-bmitch** [comment]
This proposal has been approved with 7 approvals and 2 no-votes:

Approved:

* @sudo-bmitch
* @vbatts
* @giuseppe
* @estesp
* @samuelkarp
* @dmcgowan
* @cyphar

No vote:

* @rchincha
* @sajayantony

**Comment by dims** [comment]
`no-votes` ... abstension?

**Comment by kolyshkin** [comment]
Thanks everyone. I will start on implementing this in December.

**Comment by utam0k** [comment]
> @utam0k While it would be nice for us to have less of a Go-only focus in OCI, I do wonder what a user would expect from a Rust and Go implementation of some library functionality. Would they want feature parity? If so, there's a lot of ugly stuff you would need to reimplement in Rust that is mostly legacy stuff borne out of runc's history -- though hopefully we can remove some of that when we make cgroups a separate module.
> 
> How usable (as a library) do you feel is the cgroup code in youki?

Indeed. I agree with you.

**Pull request merged**
